File: packages/python/scripts/release_metadata.py
from __future__ import annotations

import os
import re
import sys
from pathlib import Path

from packaging.version import InvalidVersion, Version


def resolve(tag: str) -> dict[str, str]:
    if not tag.startswith("python-v"):
        raise ValueError("Python release tags must start with python-v.")
    tag_version = tag.removeprefix("python-v")
    manifest = Path(__file__).parents[1] / "pyproject.toml"
    match = re.search(r'^version = "([^"]+)"$', manifest.read_text(), re.MULTILINE)
    if not match:
        raise ValueError("Python package version is missing from pyproject.toml.")
    package_version = match.group(1)
    if tag_version != package_version:
        raise ValueError(
            f"Tag version {tag_version} does not match Python package version {package_version}."
        )
    try:
        parsed = Version(package_version)
    except InvalidVersion as error:
        raise ValueError(f"Invalid Python package version: {package_version}.") from error
    return {"version": package_version, "prerelease": str(parsed.is_prerelease).lower()}


def main() -> None:
    if len(sys.argv) != 3:
        raise SystemExit("usage: release_metadata.py <tag> <output-path>")
    with Path(sys.argv[2]).open("a", encoding="utf-8") as handle:
        for name, value in resolve(sys.argv[1]).items():
            handle.write(f"{name}={value}{os.linesep}")


if __name__ == "__main__":
    main()

File: packages/typescript/scripts/release-metadata.mjs
import { appendFile, readFile } from "node:fs/promises";
import { pathToFileURL } from "node:url";

const TAG_PREFIX = "typescript-v";

export const resolveReleaseMetadata = (tag, manifest) => {
  if (typeof tag !== "string" || !tag.startsWith(TAG_PREFIX) || tag.length === TAG_PREFIX.length) {
    throw new Error("Expected a tag named typescript-v<version>.");
  }

  const version = manifest?.version;
  const packageName = manifest?.name;
  if (typeof version !== "string" || version.length === 0) {
    throw new Error("package.json must define a non-empty version.");
  }
  if (typeof packageName !== "string" || packageName.length === 0) {
    throw new Error("package.json must define a non-empty name.");
  }

  const tagVersion = tag.slice(TAG_PREFIX.length);
  if (tagVersion !== version) {
    throw new Error(`Tag version ${tagVersion} does not match package version ${version}.`);
  }

  const prerelease = version.includes("-");
  return {
    version,
    packageName,
    npmTag: prerelease ? "next" : "latest",
    prerelease
  };
};

const run = async () => {
  const [, , tag, outputPath] = process.argv;
  if (!tag || !outputPath) {
    throw new Error("Usage: node scripts/release-metadata.mjs <tag> <github-output-path>");
  }

  const manifest = JSON.parse(await readFile(new URL("../package.json", import.meta.url), "utf8"));
  const metadata = resolveReleaseMetadata(tag, manifest);
  await appendFile(outputPath, [
    `version=${metadata.version}`,
    `package_name=${metadata.packageName}`,
    `npm_tag=${metadata.npmTag}`,
    `prerelease=${metadata.prerelease}`,
    ""
  ].join("\n"));
};

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  await run();
}
